captum/attr/_utils/lrp_rules.py

```diff
@@ -0,0 +1,129 @@
+#!/usr/bin/env python3
+
+from abc import ABC, abstractmethod
+
+import torch
+
+
+class PropagationRule(ABC):
+    """
+    Base class for all propagation rule classes, also called Z-Rule.
+    STABILITY_FACTOR is used to assure that no zero divison occurs.
+    """
+
+    STABILITY_FACTOR = 1e-9
+
+    def forward_hook(self, module, inputs, outputs):
+        """Register backward hooks on input and output
+        tensors of linear layers in the model."""
+        self._handle_input_hooks = []
+        for input in inputs:
+            if not hasattr(input, "hook_registered"):
+                input_hook = self._create_backward_hook_input(input.data)
```

Comment:
nit: could we actually check
```
input._backward_hooks
```
instead?
https://github.com/pytorch/pytorch/blob/master/torch/tensor.py#L223

Reply:
I think the input tensor might have an output hook because it may be input of the next layer. Therefore only checking `input._backward_hooks` if it is None does not work. One would need to check if it has an input hook registered.